src/training/ea/lipizzaner_trainer.py

```diff
@@ -209,40 +217,77 @@ def _evaluate(self, generators: Population, discriminators: Population):
         fitness_labels = torch.squeeze(fitness_labels)
 
         # Evaluate generators
-        evaluate_fitness(
+        self.evaluate_fitness(
             attackers=generators,
             defenders=discriminators,
-            input_var=fitness_samples,
+            input_data=fitness_samples,
             fitness_mode=self.fitness_mode,
-            alpha=self.alpha,
-            beta=self.beta,
-            diverse_fitness=self.diverse_fitness,
         )
 
         # Evaluate discriminators
-        evaluate_fitness(
+        self.evaluate_fitness(
             attackers=discriminators,
             defenders=generators,
-            input_var=fitness_samples,
+            input_data=fitness_samples,
             fitness_mode=self.fitness_mode,
-            alpha=self.alpha,
-            beta=self.beta,
             labels=fitness_labels,
         )
 
+    def evaluate_fitness(
+        self,
+        attackers: Population,
+        defenders: Population,
+        input_data,
+        fitness_mode,
+        input_labels=None,
+    ):
+        # TODO verify update for MOO
+        for attacker in attackers.values():
+            attacker.fitness = 0
+            for defender in defenders.values():
+                (fitness,) = float(
+                    attacker.genome.compute_loss_against(
+                        defender.genome,
+                        input_data,
+                        labels=input_labels,
+                        alpha=self.alpha,
+                        beta=self.beta,
+                        diverse_fitness=self.diverse_fitness,
+                    )
+                )
+                if fitness_mode == "best" and fitness > attacker.fitness:
+                    attacker.fitness = fitness
+                elif fitness_mode == "worse" and fitness < attacker.fitness:
+                    attacker.fitness = fitness
+                elif fitness_mode == "average":
+                    attacker.fitness = (fitness + attacker.fitness) / len(defenders)
+
+    #
+    # SELECTION
+    #
+
     def _select(self, population):
-        return tournament_selection(
-            population,
-            self._population_size,
-            self._tournament_size,
-            self._logger,
-        )
+        return self.tournament_selection(population)
 
-    def _mutate(self, attackers: Population, defenders: Population, skip):
-        if skip:
-            return
+    def tournament_selection(self, population: Population):
```

Comment:
Same opinion I have about `evaluate_fitness`, but this will be moved to the `Selector` class anyway